### ISSUE TITLE
View attachments in browser

### ISSUE THREAD

**Issue by harmanb** [BookStackApp/BookStack#2779]
Hi, first off I wanted to thank you for making/maintaining this app, it's great. My company is planning on migrating our documentation (mostly word docs/pdf's) onto a bookstack but a major hurdle is transferring the existing documentation. It would be extremely valuable to be able to view attached documents in the app instead of having to download them. 

This request is somewhat related to issue #449 where the request was to be able to copy and paste documents with images into the editor. It was noted that that feature had technical challenges, but having either the copying feature or the viewing in browser feature would open Bookstack up to potential users that have a large amount of existing documentation that would be difficult to migrate over.

**Comment by ssddanbrown** [comment]
Thanks for your feedback @harmanb. I think this request quite significantly overlaps with existing issues #1464 and #705 therefore I'm going to close this off.

If this request was for a more significant preview "Popup/Overlay" interface for various file types, then that's really something that I feel would be out of scope for us at this time due to the effort required while somewhat working against the core content focus of BookStack.

**Comment by harmanb** [comment]
Hi @ssddanbrown thank you for the response. What we were hoping for at my organization was a way to have documents we already have created viewable/searchable within our BookStack instance. This is because we have hundreds of files of existing documentation serving as a barrier to adoption. 

After some digging we may have found a solution that could achieve this and would also open up BookStack to other potential users that have existing documentation that would be hard to port over. 

We found a js plugin that converts .docx files to html. Using the plugin it would be possible to simply append the result of its conversion to the contents of the wysiwyg editor. This would be a valuable feature because it allows users to upload their existing documentation as Pages, thus making it searchable, editable and viewable in the browser while maintaining formatting. 

I would appreciate your feedback on this feature and if you foresee any issues with it. Thanks. 

Here is a link to the plugin: https://github.com/mwilliamson/mammoth.js/
Here is a link to a demo: https://mike.zwobble.org/projects/mammoth/

**Comment by ssddanbrown** [comment]
Hi @harmanb,
Data importing is complex and highly variable, while being something that won't provide daily ongoing benefit to everyday BookStack users hence it not something I'd be looking to directly support as a feature within the core platform.

That said, these kind of tasks are well-suited to the API that we have. 

Mammoth looks good, providing a very clean output. The main issue will be correctly dealing with images. We don't have a proper way to handle images via the API yet, although #2700 has been put forward but I've not yet had a chance to properly review this. 

There might be some kind of hack for the mean time, manually organising the files and generating the database image entries. I'll have a think.

**Comment by harmanb** [comment]
Thank you for the response @ssddanbrown. I'll definitely look in to the API and see if I can get something working for my use case with that.

**Comment by MStan16** [comment]
Hi @harmanb - Any success with this?  I am trying to do similar thing with uploading word, powerpoint, excel, pdf, etc files to Bookstack, and rather have to download the file from Bookstack to view, would like to be able to view the file within Bookstack.

**Comment by ssddanbrown** [comment]
@harmanb Issue #2700 has now been merged which supports base64 images for easier conversion and importing.

With our API script examples repo, I've added an example using the mammoth library you suggested in combination with the BookStack API: https://github.com/BookStackApp/api-scripts/tree/main/node-docx-to-page

I've tested with a few sample documents, seems to work fairly well. The script is all commented and designed to be simple for readability. Hopefully that might help you put something together.
